File: find_all_anagrams_in_a_string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        from collections import defaultdict
        
        # output array
        anagram_index = []

        # if p (small string) > s
        if (len(p)>len(s)):
            return anagram_index

        # finding small string map
        small_string_map = defaultdict(int)
        for x in p:
            small_string_map[x] += 1

        # window slider subarray map
        substring_map = defaultdict(int)

        # window slider base case
        for i in range(len(p)):
            substring_map[s[i]] += 1

        # check if anagram for base case
        if substring_map == small_string_map:
            anagram_index.append(0)

        # other cases: looping from 1
        for i in range( 1, len(s)-len(p)+1 ):

            # deleting left most character from substring map
            left_most_char = s[i-1]
            substring_map[left_most_char] -= 1
            # deleting if left most key is empty value = 0
            if substring_map[left_most_char] == 0:
                del substring_map[left_most_char]

            # adding right most character to substring map
            right_most_char = s[i+len(p)-1]
            substring_map[right_most_char] += 1       

            # check if anagram
            if (substring_map == small_string_map):
                anagram_index.append(i)
        return anagram_index

File: find_all_anagrams_in_a_string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    let anagram_index = [];

    // anagram checker function
    const checkAnagram = (p_dict,s_window_dict, anagram_index, position) => {
        let temp = true;
        const p_keys = Object.keys(p_dict);
        const s_keys = Object.keys(s_window_dict);
        if(p_keys.length == s_keys.length){
            for (const key of p_keys){
                if (p_dict[key] !== s_window_dict[key]){
                    continue;
                    temp = false;
                }
            }
            if(temp == true){
                anagram_index = [...anagram_index,position];
                        }
        }
        console.log(anagram_index);
        return anagram_index;
    }

    if (s.length < p.length){
        return anagram_index;
    }

    // finding small string/p map
    p_dict = {};
    for (const letter of p){
        if (p_dict[letter]){
            p_dict[letter]++;
        }
        else{
            p_dict[letter] = 1;
        }
    }

    // window sliding
    s_window_dict = {}
    // base case
    for (let i=0; i<p.length; i++){
        if (s_window_dict[p[i]]){
            s_window_dict[p[i]]++;
        }
        else{
            s_window_dict[p[i]] = 1;
        }
    }
    // if anagram for base case
    anagram_index = checkAnagram(p_dict,s_window_dict,anagram_index,0);

    // window slider cases
    for (let i=1; i< s.length-p.length + 1; i++){

        // deleting left char
        const left_char = s[i-1];
        s_window_dict[left_char]--;

        if (s_window_dict[left_char]==0){
            delete s_window_dict.left_char;
        }       

        // adding right char
        const right_char = s[i+p.length-1];
        if (s_window_dict[right_char]){
            s_window_dict[right_char]++;
        }
        else{
            s_window_dict[right_char] = 0;
        }

        // if anagram for window slider cases
        anagram_index = checkAnagram(p_dict,s_window_dict,anagram_index,i);
    }

};

console.log(findAnagrams('cbaebabacd','abc'));